File: benches/comparison_with_other_libs/pteros/within_size_bench.py
from pteros import *
import time
import numpy as np

def within_bench(sys,n_res,d):
    for start_res in range(0,100,10):
        sel = sys(f"within {d} of resid {start_res}-{start_res+n_res}")


def within_pbc_bench(sys,n_res,d):
    for start_res in range(0,100,10):
        sel = sys(f"within {d} pbc of resid {start_res}-{start_res+n_res}")


sys=System("tests/albumin.pdb")
N = 1

def run(bench,sys,n_res,d,out):
    times = np.zeros(N)
    for i in range(N):
        t = time.process_time()
        bench(sys,n_res,d)
        times[i] = time.process_time() - t
    print(f"{n_res} {d} {times.mean()}±{times.std()}")
    out.write(f"{d} {times.mean()}\n")


for n_res in [1,20,40,60]:
    with open(f"target/pteros_{n_res}.dat","w") as out:
        for d in [0.3+0.1*val for val in range(40)]:
            run(within_bench,sys,n_res,d,out)
    
    with open(f"target/pteros_pbc_{n_res}.dat","w") as out:
        for d in [0.3+0.1*val for val in range(40)]:
            run(within_pbc_bench,sys,n_res,d,out)


# for d in [0.3+0.1*val for val in range(50)]:
#     t = time.process_time()
#     sel = sys(f"within {d} of resindex 0-200")
#     elapsed = time.process_time() - t
#     print(f"{d} {elapsed} {sel.size()}")

File: benches/within_size_bench.rs
#[cfg(test)]
mod tests {
    use molar::core::Source;
    use std::io::prelude::*;
    use std::{fs::File, io::BufWriter};

    fn run(mut code: impl FnMut()) -> f32 {
        let t = std::time::Instant::now();
        code();
        t.elapsed().as_secs_f32()
    }

    #[test]
    fn within_size_benchmark() -> anyhow::Result<()> {
        let src = Source::serial_from_file("tests/albumin.pdb")?;

        for pbc in vec!["",] {
        //for pbc in vec!["","pbc"] {
            let pbc_str = if pbc != "" {
                "pbc"
            } else {
                ""
            };

            for n_res in vec![1, 20, 40, 60] {
                let mut out =
                    BufWriter::new(File::create(format!("target/molar2_ref_{}{}.dat", pbc, n_res))?);
                for i in 0..40 {
                    let d = 0.3 + 0.1 * i as f32;
                    let t = run(|| {
                        for start_res in (0..100).step_by(10) {
                            src.select_str(format!(
                                "within {} {} of resid {}:{}",
                                d,
                                pbc_str,
                                start_res,
                                start_res + n_res
                            ))
                            .unwrap();
                        }
                    });
                    writeln!(out, "{d} {t}")?;
                    println!("{d} {t}");
                }
            }
        }

        Ok(())
    }
}
